Settings > Collection field type Numeric, can't edit Length of decimals/doubles

Hi guys,

Today I ran into the bug which @rijkvanzanten and I already discussed on Discord. This will be the reminder of the bug.

The actual bug:
We can't edit the field length in here, if the MySQL databasetype === double.
![image](https://user-images.githubusercontent.com/48241027/95263640-5f943b80-082e-11eb-9b93-2c737398e9c4.png)

`error.js:9 Error: SQLSTATE[42000]: Syntax error or access violation: 1439 Display width out of range for column 'test' (max = 255)`

The standard length of `10` has to be used always. 

I ran into this issue when I was working with longitudes and latitudes. Workaround for me was to parse them to strings, and when using them in code again, parse them back to doubles.

--- a/api/package.json
+++ b/api/package.json
@@ -102,7 +102,7 @@
 		"jsonwebtoken": "^8.5.1",
 		"keyv": "^4.0.1",
 		"knex": "^0.21.4",
-		"knex-schema-inspector": "0.0.20",
+		"knex-schema-inspector": "^0.0.21",
 		"liquidjs": "^9.14.1",
 		"lodash": "^4.17.19",
 		"macos-release": "^2.4.1",

--- a/api/src/services/fields.ts
+++ b/api/src/services/fields.ts
@@ -59,7 +59,7 @@ export class FieldsService {
 			const data = {
 				collection: column.table,
 				field: column.name,
-				type: column ? getLocalType(column.type, field?.special) : 'alias',
+				type: column ? getLocalType(column, field) : 'alias',
 				schema: column,
 				meta: field || null,
 			};
@@ -171,7 +171,7 @@ export class FieldsService {
 		const data = {
 			collection,
 			field,
-			type: column ? getLocalType(column.type, fieldInfo?.special) : 'alias',
+			type: column ? getLocalType(column, fieldInfo) : 'alias',
 			meta: fieldInfo || null,
 			schema: column || null,
 		};
@@ -248,8 +248,11 @@ export class FieldsService {
 					);
 				} else if (['float', 'decimal'].includes(field.type)) {
 					const type = field.type as 'float' | 'decimal';
-					/** @todo add precision and scale support */
-					column = table[type](field.field /* precision, scale */);
+					column = table[type](
+						field.field,
+						field.schema?.precision || 10,
+						field.schema?.scale || 5
+					);
 				} else if (field.type === 'csv') {
 					column = table.string(field.field);
 				} else {

--- a/api/src/services/payload.ts
+++ b/api/src/services/payload.ts
@@ -236,7 +236,7 @@ export class PayloadService {
 
 		const columnsWithType = columnsInCollection.map((column) => ({
 			name: column.name,
-			type: getLocalType(column.type),
+			type: getLocalType(column),
 		}));
 
 		const dateColumns = columnsWithType.filter((column) =>

--- a/api/src/utils/get-default-value.ts
+++ b/api/src/utils/get-default-value.ts
@@ -2,7 +2,7 @@ import { Column } from 'knex-schema-inspector/dist/types/column';
 import getLocalType from './get-local-type';
 
 export default function getDefaultValue(column: Column) {
-	const type = getLocalType(column.type);
+	const type = getLocalType(column);
 
 	let defaultValue = column.default_value || null;
 

--- a/api/src/utils/get-local-type.ts
+++ b/api/src/utils/get-local-type.ts
@@ -1,4 +1,5 @@
-import { types } from '../types';
+import { Column } from 'knex-schema-inspector/lib/types/column';
+import { FieldMeta, types } from '../types';
 
 /**
  * Typemap graciously provided by @gpetrov
@@ -80,14 +81,19 @@ const localTypeMap: Record<string, { type: typeof types[number]; useTimezone?: b
 };
 
 export default function getLocalType(
-	databaseType: string,
-	special?: string[] | null
+	column: Column,
+	field?: FieldMeta
 ): typeof types[number] | 'unknown' {
-	const type = localTypeMap[databaseType.toLowerCase().split('(')[0]];
+	const type = localTypeMap[column.type.toLowerCase().split('(')[0]];
 
-	if (special?.includes('json')) return 'json';
-	if (special?.includes('csv')) return 'csv';
-	if (special?.includes('uuid')) return 'uuid';
+	/** Handle Postgres numeric decimals */
+	if (column.type === 'numeric' && column.precision !== null && column.scale !== null) {
+		return 'decimal';
+	}
+
+	if (field?.special?.includes('json')) return 'json';
+	if (field?.special?.includes('csv')) return 'csv';
+	if (field?.special?.includes('uuid')) return 'uuid';
 
 	if (type) {
 		return type.type;

--- a/app/src/lang/en-US/index.json
+++ b/app/src/lang/en-US/index.json
@@ -154,6 +154,7 @@
 	"focal_length": "Focal Length",
 
 	"schema_setup_title": "How should this field save to the database?",
+	"schema_field_title": "How should this field behave in the app?",
 
 	"create_field": "Create Field",
 	"update_field": "Update Field",
@@ -263,6 +264,7 @@
 	"enter_a_value": "Enter a value...",
 	"enter_a_placeholder": "Enter a placeholder...",
 	"length": "Length",
+	"precision_scale": "Precision & Scale",
 	"readonly": "Readonly",
 	"hide_on_detail": "Hide on Detail",
 	"hide_on_browse": "Hide on Browse",
@@ -758,6 +760,8 @@
 	"collection_setup": "Collection Setup",
 	"optional_system_fields": "Optional System Fields",
 
+	"value_unique": "Value has to be unique",
+
 	"all_activity": "All Activity",
 
 	"badge": "Badge",
@@ -1175,8 +1179,9 @@
   "field_already_exists": "Field Already Exists: {field}",
   "field_created": "Field Created",
   "field_removed": "Field Removed",
-  "field_setup_interface": "What type of field are you creating?",
   "field_setup_schema": "How should it store content in the database?",
+  "field_setup_field": "How should it store content in the database?",
+  "field_setup_interface": "What type of field are you creating?",
   "field_setup_relation": "How is the relationship's data stored?",
   "field_setup_options": "All set! Just review these interface options...",
   "field_type": "Field Type",

--- a/app/src/modules/settings/routes/data-model/field-detail/components/field.vue
+++ b/app/src/modules/settings/routes/data-model/field-detail/components/field.vue
@@ -1,0 +1,106 @@
+<template>
+	<div>
+		<h2 class="type-title">{{ $t('schema_field_title') }}</h2>
+
+		<div class="form">
+			<div class="field half-left" v-if="fieldData.meta">
+				<div class="label type-label">{{ $t('readonly') }}</div>
+				<v-checkbox v-model="fieldData.meta.readonly" :label="$t('disabled_editing_value')" block />
+			</div>
+
+			<div class="field half-right" v-if="fieldData.meta">
+				<div class="label type-label">{{ $t('hidden') }}</div>
+				<v-checkbox v-model="fieldData.meta.hidden" :label="$t('hidden_on_detail')" block />
+			</div>
+
+			<div class="field full">
+				<div class="label type-label">{{ $t('note') }}</div>
+				<v-input v-model="fieldData.meta.note" :placeholder="$t('add_note')" />
+			</div>
+
+			<div class="field full">
+				<div class="label type-label">{{ $t('translations') }}</div>
+				<interface-repeater
+					v-model="fieldData.meta.translations"
+					:template="'{{ translation }} ({{ language }})'"
+					:fields="[
+						{
+							field: 'language',
+							type: 'string',
+							name: $t('language'),
+							meta: {
+								interface: 'system-language',
+								width: 'half',
+							},
+							schema: {
+								default_value: 'en-US',
+							},
+						},
+						{
+							field: 'translation',
+							type: 'string',
+							name: $t('translation'),
+							meta: {
+								interface: 'text-input',
+								width: 'half',
+								options: {
+									placeholder: 'Enter a translation...',
+								},
+							},
+						},
+					]"
+				/>
+			</div>
+		</div>
+	</div>
+</template>
+
+<script lang="ts">
+import { defineComponent, computed } from '@vue/composition-api';
+import useSync from '@/composables/use-sync';
+import { types } from '@/types';
+import i18n from '@/lang';
+import { state } from '../store';
+
+export default defineComponent({
+	props: {
+		isExisting: {
+			type: Boolean,
+			required: true,
+		},
+		type: {
+			type: String,
+			required: true,
+		},
+	},
+	setup(props, { emit }) {
+		return {
+			fieldData: state.fieldData,
+		};
+	},
+});
+</script>
+
+<style lang="scss" scoped>
+@import '@/styles/mixins/breakpoint';
+@import '@/styles/mixins/form-grid';
+
+.type-title {
+	margin-bottom: 32px;
+}
+
+.form {
+	--v-form-vertical-gap: 32px;
+	--v-form-horizontal-gap: 32px;
+
+	@include form-grid;
+}
+
+.monospace {
+	--v-input-font-family: var(--family-monospace);
+}
+
+.required {
+	--v-icon-color: var(--primary);
+}
+</style>

--- a/app/src/modules/settings/routes/data-model/field-detail/components/schema.vue
+++ b/app/src/modules/settings/routes/data-model/field-detail/components/schema.vue
@@ -18,7 +18,7 @@
 				/>
 			</div>
 
-			<div class="field">
+			<div class="field half">
 				<div class="label type-label">
 					{{ $t('type') }}
 					<v-icon class="required" sup name="star" />
@@ -34,12 +34,54 @@
 				/>
 			</div>
 
-			<div class="field full">
-				<div class="label type-label">{{ $t('note') }}</div>
-				<v-input v-model="fieldData.meta.note" :placeholder="$t('add_note')" />
-			</div>
+			<template v-if="['decimal', 'float'].includes(fieldData.type) === false">
+				<div class="field half" v-if="fieldData.schema">
+					<div class="label type-label">{{ $t('length') }}</div>
+					<v-input
+						type="number"
+						:placeholder="fieldData.type !== 'string' ? $t('not_available_for_type') : '255'"
+						:disabled="isExisting || fieldData.type !== 'string'"
+						v-model="fieldData.schema.max_length"
+					/>
+				</div>
+			</template>
 
-			<div class="field full" v-if="fieldData.schema && fieldData.schema.is_primary_key === false">
+			<template v-else>
+				<div class="field half" v-if="fieldData.schema">
+					<div class="label type-label">{{ $t('precision_scale') }}</div>
+					<div class="precision-scale">
+						<v-input type="number" :placeholder="10" v-model="fieldData.schema.precision" />
+
+						<v-input type="number" :placeholder="5" v-model="fieldData.schema.scale" />
+					</div>
+				</div>
+			</template>
+
+			<template v-if="['uuid', 'date', 'time', 'datetime', 'timestamp'].includes(fieldData.type)">
+				<div class="field half-left">
+					<div class="label type-label">{{ $t('on_create') }}</div>
+					<v-select :items="onCreateOptions" v-model="onCreateValue" />
+				</div>
+
+				<div class="field half-right">
+					<div class="label type-label">{{ $t('on_update') }}</div>
+					<v-select :items="onUpdateOptions" v-model="onUpdateValue" />
+				</div>
+			</template>
+
+			<!-- @TODO see https://github.com/directus/next/issues/639
+
+			<div class="field half-left" v-if="fieldData.schema">
+				<div class="label type-label">{{ $t('unique') }}</div>
+				<v-checkbox
+					:label="$t('value_unique')"
+					:input-value="fieldData.schema.is_unique === false"
+					@change="fieldData.schema.is_unique = !$event"
+					block
+				/>
+			</div> -->
+
+			<div class="field full" v-if="fieldData.schema && fieldData.schema.is_primary_key !== true">
 				<div class="label type-label">{{ $t('default_value') }}</div>
 				<v-input
 					v-if="['string', 'uuid'].includes(fieldData.type)"
@@ -82,29 +124,7 @@
 				/>
 			</div>
 
-			<template v-if="['uuid', 'date', 'time', 'datetime', 'timestamp'].includes(fieldData.type)">
-				<div class="field">
-					<div class="label type-label">{{ $t('on_create') }}</div>
-					<v-select :items="onCreateOptions" v-model="onCreateValue" />
-				</div>
-
-				<div class="field">
-					<div class="label type-label">{{ $t('on_update') }}</div>
-					<v-select :items="onUpdateOptions" v-model="onUpdateValue" />
-				</div>
-			</template>
-
-			<div class="field" v-if="fieldData.schema">
-				<div class="label type-label">{{ $t('length') }}</div>
-				<v-input
-					type="number"
-					:placeholder="fieldData.type !== 'string' ? $t('not_available_for_type') : '255'"
-					:disabled="isExisting || fieldData.type !== 'string'"
-					v-model="fieldData.schema.max_length"
-				/>
-			</div>
-
-			<div class="field" v-if="fieldData.schema">
+			<div class="field half-left" v-if="fieldData.schema">
 				<div class="label type-label">{{ $t('required') }}</div>
 				<v-checkbox
 					:input-value="fieldData.schema.is_nullable === false"
@@ -113,58 +133,6 @@
 					block
 				/>
 			</div>
-
-			<div class="field" v-if="fieldData.meta">
-				<div class="label type-label">{{ $t('readonly') }}</div>
-				<v-checkbox v-model="fieldData.meta.readonly" :label="$t('disabled_editing_value')" block />
-			</div>
-
-			<div class="field" v-if="fieldData.meta">
-				<div class="label type-label">{{ $t('hidden') }}</div>
-				<v-checkbox v-model="fieldData.meta.hidden" :label="$t('hidden_on_detail')" block />
-			</div>
-
-			<div class="field full">
-				<div class="label type-label">{{ $t('translations') }}</div>
-				<interface-repeater
-					v-model="fieldData.meta.translations"
-					:template="'{{ translation }} ({{ language }})'"
-					:fields="[
-						{
-							field: 'language',
-							type: 'string',
-							name: $t('language'),
-							meta: {
-								interface: 'system-language',
-								width: 'half',
-							},
-							schema: {
-								default_value: 'en-US',
-							},
-						},
-						{
-							field: 'translation',
-							type: 'string',
-							name: $t('translation'),
-							meta: {
-								interface: 'text-input',
-								width: 'half',
-								options: {
-									placeholder: 'Enter a translation...',
-								},
-							},
-						},
-					]"
-				/>
-			</div>
-
-			<!--
-			@todo add unique when the API supports it
-
-			<div class="field">
-				<div class="label type-label">{{ $t('unique') }}</div>
-				<v-input v-model="fieldData.schema.unique" />
-			</div> -->
 		</div>
 	</div>
 </template>
@@ -422,35 +390,17 @@ export default defineComponent({
 
 <style lang="scss" scoped>
 @import '@/styles/mixins/breakpoint';
+@import '@/styles/mixins/form-grid';
 
 .type-title {
 	margin-bottom: 32px;
 }
 
 .form {
-	display: grid;
-	grid-gap: 32px;
-	grid-template-columns: 1fr 1fr;
-}
+	--v-form-vertical-gap: 32px;
+	--v-form-horizontal-gap: 32px;
 
-.field {
-	grid-column: 1 / span 2;
-
-	@include breakpoint(small) {
-		grid-column: auto;
-	}
-}
-
-.full {
-	grid-column: 1 / span 2;
-
-	@include breakpoint(small) {
-		grid-column: 1 / span 2;
-	}
-}
-
-.label {
-	margin-bottom: 8px;
+	@include form-grid;
 }
 
 .monospace {
@@ -459,5 +409,11 @@ export default defineComponent({
 
 .required {
 	--v-icon-color: var(--primary);
+}
+
+.precision-scale {
+	display: grid;
+	grid-gap: 12px;
+	grid-template-columns: 1fr 1fr;
 }
 </style>

--- a/app/src/modules/settings/routes/data-model/field-detail/field-detail.vue
+++ b/app/src/modules/settings/routes/data-model/field-detail/field-detail.vue
@@ -50,6 +50,13 @@
 			:type="localType"
 		/>
 
+		<setup-field
+			v-if="currentTab[0] === 'field'"
+			:is-existing="field !== '+'"
+			:collection="collection"
+			:type="localType"
+		/>
+
 		<setup-relationship
 			v-if="currentTab[0] === 'relationship'"
 			:is-existing="field !== '+'"
@@ -97,6 +104,7 @@ import { defineComponent, onMounted, ref, computed, reactive, PropType, watch, t
 import SetupTabs from './components/tabs.vue';
 import SetupActions from './components/actions.vue';
 import SetupSchema from './components/schema.vue';
+import SetupField from './components/field.vue';
 import SetupRelationship from './components/relationship.vue';
 import SetupTranslations from './components/translations.vue';
 import SetupInterface from './components/interface.vue';
@@ -119,6 +127,7 @@ export default defineComponent({
 		SetupTabs,
 		SetupActions,
 		SetupSchema,
+		SetupField,
 		SetupRelationship,
 		SetupTranslations,
 		SetupInterface,
@@ -195,6 +204,11 @@ export default defineComponent({
 						text: i18n.t('schema'),
 						value: 'schema',
 						disabled: false,
+					},
+					{
+						text: i18n.tc('field', 1),
+						value: 'field',
+						disabled: interfaceDisplayDisabled(),
 					},
 					{
 						text: i18n.t('interface'),

--- a/app/src/modules/settings/routes/data-model/field-detail/store.ts
+++ b/app/src/modules/settings/routes/data-model/field-detail/store.ts
@@ -37,6 +37,8 @@ function initLocalStore(collection: string, field: string, type: typeof localTyp
 				default_value: undefined,
 				max_length: undefined,
 				is_nullable: true,
+				precision: null,
+				scale: null,
 			},
 			meta: {
 				hidden: false,

